packages/legend-query/src/stores/QueryBuilderPropertyEditorState.ts

```diff
@@ -39,21 +40,339 @@ import {
+ */
+const checkForOverridingMilestoningQuery = (
+  derivedPropertyExpressionState: QueryBuilderDerivedPropertyExpressionState,
+  targetStereotype: MILESTONING_STEREOTYPE,
+): boolean => {
+  const sourceStereotype = getDerivedPropertyMilestoningSteoreotype(
+    derivedPropertyExpressionState.derivedProperty,
+    derivedPropertyExpressionState.queryBuilderState.graphManagerState.graph,
+  );
+  if (
+    sourceStereotype === MILESTONING_STEREOTYPE.PROCESSING_TEMPORAL &&
+    targetStereotype === MILESTONING_STEREOTYPE.BITEMPORAL &&
+    derivedPropertyExpressionState.parameterValues.length === 1
+  ) {
+    return true;
+  } else {
+    return false;
+  }
+};
+
+/**
+ * Check if the parameter value of the milestoned property is
+ * the same as those specified in global scope, so that we can
+ * potentially replace them with propgated value.
+ */
+const checkForMilestoningParameterEquality = (
+  stereotype: MILESTONING_STEREOTYPE,
+  idx: number,
+  milestoningParameter: ValueSpecification,
+  querySetupState: QueryBuilderSetupState,
+): boolean => {
+  const checkIfEquivalent = (
+    param1: ValueSpecification | undefined,
+    param2: ValueSpecification | undefined,
+  ): boolean => {
+    if (
+      param1 instanceof VariableExpression &&
+      param2 instanceof VariableExpression
+    ) {
+      return param1.name === param2.name;
+    } else if (
+      param1 instanceof PrimitiveInstanceValue &&
+      param2 instanceof PrimitiveInstanceValue
+    ) {
+      if (
+        param1.genericType.value.rawType.name === PRIMITIVE_TYPE.LATESTDATE &&
+        param2.genericType.value.rawType.name === PRIMITIVE_TYPE.LATESTDATE
+      ) {
+        return true;
+      } else if (
+        param1.genericType.value.rawType.name ===
+        param2.genericType.value.rawType.name
+      ) {
+        return param1.values[0] === param2.values[0];
+      }
+      return false;
+    }
+    return false;
+  };
+  switch (stereotype) {
+    case MILESTONING_STEREOTYPE.BITEMPORAL:
+      if (
+        idx === 0 &&
+        checkIfEquivalent(milestoningParameter, querySetupState.processingDate)
+      ) {
+        return true;
+      } else if (
+        idx === 1 &&
+        checkIfEquivalent(milestoningParameter, querySetupState.businessDate)
+      ) {
+        return true;
+      } else {
+        return false;
+      }
+    case MILESTONING_STEREOTYPE.PROCESSING_TEMPORAL:
+      if (
+        checkIfEquivalent(milestoningParameter, querySetupState.processingDate)
+      ) {
+        return true;
+      } else {
+        return false;
+      }
+    case MILESTONING_STEREOTYPE.BUSINESS_TEMPORAL:
+      if (
+        checkIfEquivalent(milestoningParameter, querySetupState.businessDate)
+      ) {
+        return true;
+      } else {
+        return false;
+      }
+    default:
+  }
+  return false;
+};
+
+export const isDefaultDatePropagationSupported = (
```

Comment:
Let's document this method. And I meant a serious doc since this method seems to capture a fairly complex logic which you use diagram to explain in your PR. It might be a good idea to try to chart this out similar to what we did in `ChangeDetectionState`